xarray/core/dataset.py

```diff
@@ -9026,7 +9025,25 @@ def polyfit(
         variables = {}
         skipna_da = skipna
 
-        x = get_clean_interp_index(self, dim, strict=False)
+        x: Any = self.coords[dim].to_index()
+        # Special case for non-standard calendar indexes
+        # Numerical datetime values are defined with respect to 1970-01-01T00:00:00 in units of nanoseconds.
+        if isinstance(x, CFTimeIndex | pd.DatetimeIndex):
+            offset = type(x[0])(1970, 1, 1)
+            if isinstance(x, CFTimeIndex):
+                x = x.values
+            x = Variable(
+                data=datetime_to_numeric(x, offset=offset, datetime_unit="ns"),
+                dims=(dim,),
+            )
+
+        try:
+            x = x.values.astype(np.float64)
+        except TypeError:
+            raise TypeError(
+                f"Dim {dim!r} must be castable to float64, got {type(x).__name__}."
+            )
+
```

Comment:
it's ugly but lets use `core.missing._floatize_x` to at  least use the same logic everywhere.

Reply:
Done :)